odoo/addons/base/models/res_lang.py

```diff
@@ -155,7 +169,190 @@ def fix_datetime_format(format):
+            _logger.info("Skip translations extraction, module %s is in state %s", addons, module.state)
+            return
+
+        src_file = os.path.join(tmp, extracted_file)
+        with open(src_file, 'r') as po_file:
+            po_content = base64.b64encode(po_file.read().encode()).decode()
+
+        # sudo to be able to search and write, even if no rights on ir.module.module
+        Attachment = self.env['ir.attachment'].sudo()
+
+        attached_po = Attachment.search([
+            ('res_model', '=', 'ir.module.module'),
+            ('res_id', '=', module.id)
+            ('name', '=', extracted_file)
+        ], limit=1)
+        if attached_po:
+            attached_po.write({'datas': po_content})
+        else:
+            attached_po = Attachment.create({
+                'datas': po_content,
+                'res_model': 'ir.module.module',
+                'res_id': module.id,
+                'name': extracted_file,
+            })
+        return attached_po
+
+    @api.model
+    def _extract_i18n_file_content(self, fileobj, lang, module_list, extraction_method):
+        """ Extract the translations from the given archive
+
+        The expected archive is a .tar.xz file using the structure:
+            <module>/i18n/<lang>.po
+
+        Regional variants are accepted (e.g. fr.tar.xz can contain both fr and
+        fr_BE files)
+
+        :param fileobj: a file object containing the compressed translation files
+        :param lang: the simplified language code to load (e.g. 'fr')
+        :param module_list: the list of modules to update from this archive
+        :param extraction_method: the method to call with the extracted .po file
+        """
+        with tarfile.open(mode='r:xz', fileobj=fileobj) as tar_content:
+            with tempfile.TemporaryDirectory() as tmp:
+                for filename in tar_content.getnames():
+                    if not filename.endswith('.po'):
+                        _logger.info("Skip unexpected file %s", filename)
+                        continue
+
+                    # TODO different separators for windows in tar?
+                    addons = filename.split('/')[0]
+                    if addons not in module_list:
+                        _logger.debug("Skip translations for unexpected module %s", addons)
+                        continue
+
+                    po_lang = filename.split('/')[-1][:-3]
+                    if po_lang.split('_')[0] != lang:
+                        _logger.debug("Skip translations for unexpected language %s", po_lang)
+                        continue
+
+                    if addons not in http.addons_manifest:
+                        _logger.debug("Skip translations for module %s, unknown addons-path", addons)
+                        continue
+
+                    _logger.debug("Extracting translation file %s to %s", filename, tmp)
+                    tar_content.extract(filename, path=tmp)
```

Comment:
Might be able to skip the tempdir with `extractfile` (which returns a pseudo-file object backed by the tarchive)

Reply:
the thing is that my file is something like `account/i18n/fr.po`, having one dir made it quite easier